### DISCUSSION TITLE
Should we replace Yup with Zod?

### DISCUSSION THREAD

**Discussion by hanzelkatomas** [colinhacks/zod#1970] [Q&A] [answered]
Hey, 
we are thinking of replacing Yup with Zod. We are unhappy with few things on yup. Most of our app are multistep and kinda complicated forms, where we need to reference to another value in form values and make validation against that value. 

The problem with Yup is that it can see only on the same level in tree structure and can't see one or more levels up and this complicates our already complicated form validation. Is this solved in Zod and I can reference for any value in "object of values" of form. 
Also Yup seems to not be active lately and we don't want to build our new project on something, which is already outdated.

I will be happy for any answers. Thanks

**Comment by JacobWeisenburger** [comment, marked as answer]
Are you able to give a code example?

**Reply by hanzelkatomas** [reply]
Sure, here is code of two fields of yup validation.
```Typescript
isNew: boolean().required(),
participantsRequirements: array().of(
    object().shape({
      amount: moneyValidator().required(),
      requirementsDate: date()
        .when("isNew", {
            /* Here comes the problem, yup can't see isNew because "requirementsDate" is nested in object of array
              and yup can see only on the object level (so only property "amount").
              This can be solved via custom test() function where you can get to value which is higher using kinda hacky way.
            */
          is: (isNew: boolean) => isNew,
          then: date().min(DateTime.now().toJSDate()),
        })
    }),
  ),
  ```

**Reply by JacobWeisenburger** [reply]
I can't find a good way to solve this. But to be fair, this seems like a problem that would be difficult to solve no matter which validation library you are using.

**Reply by hanzelkatomas** [reply]
It seems to be like that. Thanks for the help.

**Reply by JacobWeisenburger** [reply]
perhaps you can move `isNew` inside the object array. that might make it easier to work with.

**Reply by hanzelkatomas** [reply]
Yep, that's one of solution.
When we want to have access to all the other field, we use the `when()` function on the the top. That's okay, but our forms are much more complicated than this and when we have to cover multiple cases, we have to chain multiple when on when, so we end up with 3 whens chained. where 90 % of code of each when is the same. This duplicates code a lot, it's not readable and very unaesthetic.
It could be much more readable if we could use `when()` inside object and it could acess all the values from form.

Also, in this case we don't want to move `isNew` to object in array. Because `isNew` represent only one field and should be on page just once, on the other hand `participantsRequirements` should be on page multiple times. Because each object in array represents value for each participant.
